searx/plugins/unit_converter.py

```diff
@@ -2,57 +2,167 @@
 
 CONVERT_KEYWORDS = ["in", "to", "as"]
 
+# inspired from https://stackoverflow.com/a/42475086
+RE_MEASURE = r'''
+(?P<sign>[-+]?)         # +/- or nothing for positive
+(\s*)                   # separator: white space or nothing
+(?P<number>[\d\.,]*)    # number: 1,000.00 (en) or 1.000,00 (de)
+(?P<E>[eE][-+]?\d+)?    # scientific notation: e(+/-)2 (*10^2)
+(\s*)                   # separator: white space or nothing
+(?P<unit>\S+)           # unit of measure
+'''
+
+ADDITIONAL_UNITS = [
+    {
+        "si_name": "Q11573",  # wikidata SI name
+        "symbol": "L",
+        "to_si": 1609.344,
+        "from_si": 1 / 1609.344,
+    },
+    {
+        "si_name": "Q11579",
+        "symbol": "°C",
+        "to_si": lambda x: x + 273.15,
+        "from_si": lambda x: x - 273.15,
+    },
+    {
+        "si_name": "Q11579",
+        "symbol": "°F",
+        "to_si": lambda x: (x + 459.67) * 5 / 9,
+        "from_si": lambda x: (x * 9 / 5) - 459.67,
+    },
+]
+
+ALIAS_SYMBOLS = {
+    '°C': ('C',),  # °C <--> C
+    '°F': ('F',),  # °F <--> F
+}
+
+
+def _parse_text_and_convert(search, from_query, to_query):
+
+    # pylint: disable=too-many-branches, too-many-locals
+
+    if not (from_query and to_query):
+        return
 
-def _convert(from_value, source_si_factor, target_si_factor):
-    return from_value * source_si_factor / target_si_factor
-
+    measured = re.match(RE_MEASURE, from_query, re.VERBOSE)
+    if not (measured and measured.group('number'), measured.group('unit')):
+        return
 
-def _parse_text_and_convert(search, splitted_query):
-    if len(splitted_query) != 2 or splitted_query[0].strip() == "" or splitted_query[1].strip() == "":
+    # filter out units which can't be normalized to a SI unit and filter out
+    # units without a symbol / arcsecond does not have a symbol
+    # https://www.wikidata.org/wiki/Q829073
+
+    symbol_to_si = []
+    for item in WIKIDATA_UNITS.values():
+        if item['to_si_factor'] and item['symbol']:
+            symbol_to_si.append(
+                (
+                    item['symbol'],
+                    item['si_name'],
+                    item['to_si_factor'],  # from_si
+                    1 / item['to_si_factor'],  # to_si
+                    item['symbol'],
+                )
+            )
+
+    for item in ADDITIONAL_UNITS:
+        symbol_to_si.append(
+            (
+                item['symbol'],
+                item['si_name'],
+                item['from_si'],
+                item['to_si'],
+                item['symbol'],
+            )
+        )
+
+    alias_items = []
+    for item in symbol_to_si:
+        for alias in ALIAS_SYMBOLS.get(item[0], ()):
+            alias_items.append(
+                (
+                    alias,
+                    item[1],
+                    item[2],  # from_si
+                    item[3],  # to_si
+                    item[0],  # origin unit
+                )
+            )
+    symbol_to_si = symbol_to_si + alias_items
```

Comment:
It would make sense to move this to an `init` function so that it doesn't run each time you search something often - don't plugins have something like that?

Or we could move it to the `fetch_wikidata_units` function in `searx.data`.

Reply:
> don't plugins have something like that?

we can define a `def init(..)` in the plugin module:

https://github.com/searxng/searxng/blob/c8d0b6529b616aa33678afa10edc414b460ea017/searx/plugins/__init__.py#L177-L181

but `plugin.init(*init_args)` is executed when the app is initialized:

https://github.com/searxng/searxng/blob/c8d0b6529b616aa33678afa10edc414b460ea017/searx/plugins/__init__.py#L228-L230

> Or we could move it to the `fetch_wikidata_units` function in `searx.data`.

Yes, makes more sense / especially if we will have lazy evaluation for it (#3443).

Reply:
> Or we could move it to the `fetch_wikidata_units` function in `searx.data`.

FYI last force push was just a rebase to get https://github.com/searxng/searxng/pull/3445

Reply:
> Or we could move it to the `fetch_wikidata_units` function in `searx.data`.

I forgot we have functions which can't stored in a JSON string:

```python
ADDITIONAL_UNITS = [
    {
        "to_si": lambda val: val + 273.15,
        "from_si": lambda val: val - 273.15,
        ...
```

I modified the code to generate the list only once and store it in a global / BTW I added documentation with more details to the module.

Reply:
is PR ready to merge?

Reply:
Yes :+1: